egg/core/trainers.py

```diff
@@ -124,44 +144,49 @@ def train_epoch(self):
         return mean_loss, mean_rest
 
     def train(self, n_epochs):
-        def _message(mode, epoch, loss, rest, as_json):
-            writer = get_summary_writer()
-            if writer:
-                writer.add_scalar(tag=f'{mode}/loss', scalar_value=loss.mean(), global_step=epoch)
-                for k, v in rest.items():
-                    writer.add_scalar(tag=f'{mode}/{k}', scalar_value=v, global_step=epoch)
-            if as_json:
-                dump = dict(mode=mode, epoch=epoch, loss=loss.mean().item())
-                for k, v in rest.items(): dump[k] = v.item() if hasattr(v, 'item') else v
-                output_message = json.dumps(dump)
-            else:
-                output_message = f'{mode}: epoch {self.epoch}, loss {loss},  {rest}'
-            print(output_message, flush=True)
-
-        while self.epoch < n_epochs:
+        for callback in self.callbacks:
+            callback.on_train_begin(self)
+
+        for epoch in range(n_epochs):
+            self.epoch = epoch  # TODO: deprecate this attribute in favor of the new callback API
+            for callback in self.callbacks:
+                callback.on_epoch_begin()
+
             train_loss, train_rest = self.train_epoch()
 
-            self.epoch += 1
+            for callback in self.callbacks:
+                callback.on_epoch_end(train_loss, train_rest)
 
+            # TODO: the logic below is to be reimplemented using the new callback API
             if self.epoch_callback:
                 self.epoch_callback(self)
-
             if self.checkpoint_freq > 0 and (self.epoch % self.checkpoint_freq == 0) and self.checkpoint_path:
                 self.save_checkpoint()
+            # TODO: the logic above is to be reimplemented using the new callback API
 
-            if self.print_train_loss:
-                _message('train', self.epoch, train_loss, train_rest, self.as_json)
-
-            if self.validation_data is not None and self.validation_freq > 0 and self.epoch % self.validation_freq == 0:
+            if self.validation_data is not None and self.validation_freq > 0 and epoch % self.validation_freq == 0:
+                for callback in self.callbacks:
+                    callback.on_test_begin()
                 validation_loss, rest = self.eval()
-                _message('validation', self.epoch, validation_loss, rest, self.as_json)
+                for callback in self.callbacks:
+                    callback.on_test_end(validation_loss, rest)
 
+                # TODO: the logic below is to be reimplemented using the new callback API
                 if self.early_stopping:
                     self.early_stopping.update_values(validation_loss, rest, train_loss, rest, self.epoch)
                     if self.early_stopping.should_stop(): break
+                # TODO: the logic above is to be reimplemented using the new callback API
+
+            if self.should_stop:
```

Comment:
Same, we could remove it until the actual logic appears; I guess at the moment it could be a bit confusing to have flag that is never set.